### ISSUE TITLE
GraphQL API (Datastore)- Field Level Auth Rule not working

### ISSUE THREAD

**Issue by SimonSch-tech** [aws-amplify/amplify-cli#6343]
Greetings,

in our project we try to define a proper GraphQL Model. Our goal is to use one model definition with different authentication types and rules for making requests via amplify datastore. We used amplifys GraphQL directives to annotate our model.

We defined the model like this:
```gql
type Entity @model
@auth(rules: [
    # Default owner access
    { allow: owner },

    # Admin group access
    { allow: groups, groups: ["Admin"] },

    # User
    { allow: groups, groups: ["User"], operations: [read] },

    # Everyone
    { allow: public, operations: [read] }
  ])
@key(name: "sort", fields: ["name"])
 {
    id: ID!
    owner: String
    name: String @auth(rules: [{ allow: owner, operations: [read] }, { allow: groups, groups: ["Admin"], operations: [read] }])
}
```

For granting non authenticated and authenticated users access to our api we implemented different authentication types by dynamically switching between
`"aws_appsync_authenticationType": "API_KEY"`
as the authentication method for unauthenticated users and
`"aws_appsync_authenticationType": "AMAZON_COGNITO_USER_POOLS"`
as the authentication method for authenticated users from cognito user pool.

Our auth rules should give owners and admins full access to CRUD operations for this type. Authenticated cognito users and unauthenticated users should only get read access to the model. For the field 'name' we want to define a specific behaviour, only the owners and users which belong to group 'Admin' should be able to read that field.

We assumed, if used field level auth rules are excluding any authentication for that specific field. So we had to grant explicit access for owners and admins. 

Our amplify library version in our ReactJS project:
`"aws-amplify": "^3.3.13",
  "aws-amplify-react": "^4.2.17",`

The expected behaviour was:
**Name field** - Unauthenticated users should not be able to read the field 'name'. Read access should only be granted to owners and group members of the cognito user pool group 'Admin.'

The actual behaviour:
**Name field** - Unauthenticated users and authenticated users, are both granted read access.

**Comment by SimonSch-tech** [comment]
@amhinson thy. Hi @edwardfoyle When can we expect first feedback for this issue?

**Comment by SwaySway** [comment]
@SimonSch-tech It looks like you are looking for a deny behavior. Currently the `@auth` rules only support allow. This is something we are considering for the next version of the transformer as well as part of the `@auth` enhancements.

RFCs: 
`@auth`: https://github.com/aws-amplify/amplify-cli/issues/1043
GraphQL vNext: https://github.com/aws-amplify/amplify-cli/issues/6217

Based on your schema if your two auth modes are `AMAZON_COGNITO_USER_POOLS` and `API_KEY` one workaround here would be to just enforce cognito authorization on the `name` field. This would prevent users signed in with `API_KEY` to see the value of `name`.

example
```gql
type Entity @model
@auth(rules: [
    { allow: owner },
    { allow: groups, groups: ["Admin"] },
    { allow: groups, groups: ["User"], operations: [read] },
    { allow: public, operations: [read] }
  ])
@key(name: "sort", fields: ["name"])
 {
    id: ID!
    owner: String
    name: String @aws_cognito_user_pools
}
```

Docs on AppSync Directives: https://docs.aws.amazon.com/appsync/latest/devguide/security-authz.html#using-additional-authorization-modes

**Comment by cfbo** [comment]
Hi there. I have a scenario that might be sort of related to this ticket.

I am implementing a "sharing" functionality.

Scenario:  
- user A saves a Data record R1. Then they can share that record R1. 
- user B access the shared R1 (I use a lambda to retrieve it due to auth issues). 
- at some point user B may want to save the "link" to R1 and I am doing this by saving that a SharedData record pointing to R1

I have a situation like that:
```
type Data
  @model
  @auth(rules: [{ allow: owner, operations: [create, update, delete, read] }]) {
  id: ID!
  content: String
  isLinkSharingEnabled: Boolean
  owner: String
}

type SharedData
  @model
  @auth(rules: [{ allow: owner, operations: [create, update, delete, read] }]) {
  id: ID!
  dataId: ID!
  # the @connection causes authorization errors
  data: Data
    @connection(fields: ["dataId"])
    @auth(rules: [{ allow: private, provider: iam }])
  owner: String
}
```
However if I use the `@connection` I get queries / mutations errors (I am using the Datastore and I am referring to the network calls made by the datastore engine). Of course if remove `@connection` the mutations are saved without errors and data retrieved correctly. But I "loose" the "FK" to that Data record which is a pain, because I would use that from the lambda when retrieving that Data record. 

The main problem is that the code generated by the Datastore (index.d.ts) for the SharedData class only contains the `data` object and not `dataId` (like in graphql), so data is always null... but would I would need is only dataId.

In other words: when I add the connection, the generated class looses the id field in favor of the object.
![image](https://user-images.githubusercontent.com/7568421/142442122-091a6e7e-a14c-46aa-8493-a40e9ed53ae8.png)

**Comment by dorontal** [comment]
@SwaySway you discuss **deny** behavior on auth fields, but even when I try to use field-level `@auth` directive to **allow** something, field-level `@auth` directives don't seem to work.

[ Note: using v2 of transformer with the latest versions of Amplify CLI (7.6.15) and library. ]

I thought you could get deny behavior on individual fields by using `@auth` on the parent model, to deny, then using `@auth` again on the individual fields to override the denial and allow only for those fields. My understanding is that model level `@auth` directives are overridden by any field level `@auth` directive - is my understanding wrong here?

Here's the schema model in which field-level `@auth` is not working. I want the owner to be able to read any field of their own `User` model, but for the specific fields that have an `@auth` directive in this schema, I want any **other user** (signed-in or via IAM-unauth role) to **not** be able to read the field-level protected fields in this schema (those with an `@auth` directive).

```graphql
type User
  @model(subscriptions: { level: off })
  @auth(
    rules: [
      { allow: owner, operations: [read] }
      { allow: private, operations: [read] }
      { allow: public, provider: iam, operations: [read] }
    ]
  ) {
  id: ID!
  colorIndex: Int!
  imageType: ImageType!
  displayName: String!
  initials: String!
  bio: String!
  place: String!
  nPlays: Int!
  nProjects: Int!
  nArrangements: Int!
  nPlaylists: Int!
  bytesQuota: Int! @auth(rules: [{ allow: owner, operations: [read] }])
  bytesUsed: Int! @auth(rules: [{ allow: owner, operations: [read] }])
  email: String! @auth(rules: [{ allow: owner, operations: [read] }])
  nUpdatesInLast24h: Int! @auth(rules: [{ allow: owner, operations: [read] }])
  nUploadsInLast24h: Int! @auth(rules: [{ allow: owner, operations: [read] }])
  totalEarned: Float! @auth(rules: [{ allow: owner, operations: [read] }])
  createdAt: AWSDateTime! @auth(rules: [{ allow: owner, operations: [read] }])
  updatedAt: AWSDateTime! @auth(rules: [{ allow: owner, operations: [read] }])
  uploadedAt: AWSDateTime! @auth(rules: [{ allow: owner, operations: [read] }])
  projects: [Project] @hasMany
}
```

The problem is that with this schema, the owner cannot view the fields that have an `@auth` directive even though the `@auth` directive says `@auth(rules: [{ allow: owner, operations: [read] }])` - either `@auth` directives don't work in this case or I misunderstood something.

The strange thing is that the owner used to be able to view those fields without any problems up till recently.

Now I get errors like this when executing the `GetUser` query by the owner to get their own user instance:
```
Caught error Error: Uncaught (in promise): Object: {"data":{"getUser":null},"errors":[{"path":["getUser","email"],"data":null,"errorType":"Unauthorized","errorInfo":null,"locations":[{"line":17,"column":5,"sourceName":null}],"message":"Not Authorized to access email on type String"},{"path":["getUser","uploadedAt"],"data":null,"errorType":"Unauthorized","errorInfo":null,"locations":[{"line":23,"column":5,"sourceName":null}],"message":"Not Authorized to access uploadedAt on type AWSDateTime"},{"path":["getUser","nUploadsInLast24h"],"data":null,"errorType":"Unauthorized","errorInfo":null,"locations":[{"line":19,"column":5,"sourceName":null}],"message":"Not Authorized to access nUploadsInLast24h on type Int"},{"path":["getUser","bytesUsed"],"data":null,"errorType":"Unauthorized","errorInfo":null,"locations":[{"line":16,"column":5,"sourceName":null}],"message":"Not Authorized to access bytesUsed on type Int"},{"path":["getUser","nUpdatesInLast24h"],"data":null,"errorType":"Unauthorized","errorInfo":null,"locations":[{"line":18,"column":5,"sourceName":null}],"message":"Not Authorized to access nUpdatesInLast24h on type Int"},{"path":["getUser","totalEarned"],"data":null,"errorType":"Unauthorized","errorInfo":null,"locations":[{"line":20,"column":5,"sourceName":null}],"message":"Not Authorized to access totalEarned on type Float"},{"path":["getUser","bytesQuota"],"data":null,"errorType":"Unauthorized","errorInfo":null,"locations":[{"line":15,"column":5,"sourceName":null}],"message":"Not Authorized to access bytesQuota on type Int"},{"path":["getUser","updatedAt"],"data":null,"errorType":"Unauthorized","errorInfo":null,"locations":[{"line":22,"column":5,"sourceName":null}],"message":"Not Authorized to access updatedAt on type AWSDateTime"},{"path":["getUser","createdAt"],"data":null,"errorType":"Unauthorized","errorInfo":null,"locations":[{"line":21,"column":5,"sourceName":null}],"message":"Not Authorized to access createdAt on type AWSDateTime"}]}
```

**IMPORTANT NOTE: ** Amplify CLI recently asked me if I would allow Auth field level directives changes to be incorporated, when I did an `amplify push` of a minor schema change recently. I said "yes". This issue was not an issue before Amplify made these changes. Is it possible that these changes are what broke field-level authorization?

**Comment by dorontal** [comment]
I'm also seeing a problem where model `@auth` is more permissive than field `@auth` but as soon as you add the field `@auth` the user **loses permission** to access **any part of the model** -- can't even read -- although the permissions allow read access at both model or field level:

This is the model `@auth`
```graphql
@auth(
    rules: [
      { allow: owner, operations: [read] }
      { allow: private, operations: [read] }
      { allow: public, provider: iam, operations: [read] }
    ]
  ) {
```
and this is the field `@auth` for the same model
```graphql
  bytesQuota: Int! @auth(rules: [{ allow: owner, operations: [read] }])
```
but as soon as you add the field rule, user can no longer access that field, or the model, even though they are allowed `read` access.

**Comment by josefaidt** [comment]
Hey y'all :wave: since this issue was created we've released a new version of the GraphQL Transformer which [changed auth rules to deny-by-default](https://docs.amplify.aws/cli/graphql/authorization-rules/). Using the following schema with Cognito User Pools and API Key as the auth mechanism we are able to allow owners to create and read their own `notes`, however when using API Key for public auth this field is not readable

```graphql
type Todo
  @model
  @auth(rules: [{ allow: owner }, { allow: public, operations: [read] }]) {
  id: ID!
  name: String!
  notes: String @auth(rules: [{ allow: owner }])
}
```

Closing issue 🙂

**Comment by dorontal** [comment]
Sorry @josefaidt - still having this issue and I've been with V2 for a long time now. This issue still stands for me and it's with V2 of the GraphQL transformer. Specificaly, it is not the schema use case you show above, but the one I show in my previous comment here - that one still does not allow the owner to read their own model, if the field `bytesQuota` is part of that query.

If you ever review what I am reporting here and see that I've done something wrong in the schema definition itself, or in my assumptions, then please let me know. Otherwise, most likely this issue has not gone away, not for me, in the use case outlined above. It's easy to try with your Todo app - have you tried that use-case with those `@auth` directive that I describe?  I know that your use case works. Mine does not. Am I making the wrong assumptions here and my use case wasn't supposed to work in the first place?  Currently it throws a bug if the owner tries to access their model in a way that retrieves the field-protected field named `bytesQuota` - that bug is an access denied bug.

**Comment by josefaidt** [comment]
Hey @dorontal I modified the example based on your update to be read-only and it appears to fetch the owner's `notes`. Are you experiencing a situation where the owners are not able to read their notes?
![fetch todos as user](https://user-images.githubusercontent.com/5033303/155623804-b49aebe8-0eae-44ed-9fa3-61ed289ce80f.png)
![fetch todos as josef](https://user-images.githubusercontent.com/5033303/155624116-aede47cb-c463-45da-b25b-5ca932ce2591.png)

**Comment by dorontal** [comment]
In some sense the answer is yes: the owner cannot read a model if the query asks for a field-level Auth rule that has owner read-only access.

But I was extrapolating from my own schema where `Todo` is a `User` model and `notes` is `bytesQuota` - and I get this error in the browser's DEV console:
```
"message":"Not Authorized to access bytesQuota on type Int"
```

**Comment by josefaidt** [comment]
Hey @dorontal would you mind filing a [separate bug report](https://github.com/aws-amplify/amplify-cli/issues/new?assignees=&labels=&template=1.bug_report.yaml) for this? I think it would be helpful to gather the additional context to better understand the root cause as I am not able to reproduce the behavior.

**Comment by dorontal** [comment]
Sure, will do.

**Comment by dorontal** [comment]
@josefaidt  I did open a new issue - but frankly I do not understand why this was necessary.

I also do not understand why this issue got closed.

@josefaidt  can you show the schema that you used for that `Todo` model -- did that schema include all the model level Auth rules that mine did?

I do not understand why this issue got closed. The problem is not solved yet. This issue describes things with enough detail.
